CIScripts/Tests/VRouterAgentTests.ps1

```diff
@@ -668,119 +716,262 @@ function Test-VRouterAgentIntegration {
+        Write-Host "        open TCP connection to $Container2Name (network: $($Network2.Name), IP: $Container2IP)"
+        Write-Host
 
-                    & docker exec $Using:Container1Name powershell -Command $Command
-                }
+        $Port = "1905"
+        $Message = "Even diamonds require polishing."
+
+        Write-Host "        Setting up TCP sender and listener on containers..."
 
-                $JobSender | Wait-Job -Timeout 5 | Out-Null
-                $JobListener | Wait-Job -Timeout 5 | Out-Null
-                $ReceivedMessage = $JobListener | Receive-Job
-                return $ReceivedMessage
+        $Functions = {
+            function New-WaitFileCommand {
+                Param ([String] $Path)
+                return "while (!(Test-Path $Path)) { Start-Sleep -Milliseconds 300 };"
             }
+        }
 
-            # TODO: Enable this test once it is actually expected to pass.
-            # .NET in microsoft/nanoserver docker image doesn't have TCPServer
-            #Write-Host "        Sent message: $Message"
-            #Write-Host "        Received message: $ReceivedMessage"
+        Invoke-Command -Session $Session2 -ScriptBlock {
 
-            #if ($Message -ne $ReceivedMessage) {
-            #    throw "Sent and received messages do not match."
-            #} else {
-            #    Write-Host "        Match!"
-            #}
+            $Functions = [ScriptBlock]::Create($Using:Functions)
+            $JobListener = Start-Job -InitializationScript $Functions -ScriptBlock {
+                param(
+                    $Port,
+                    $Container2Name,
+                    $Message)
+
+                $Command = (`
+                    ('$Server = New-Object System.Net.Sockets.TcpListener(\"0.0.0.0\", {0});' -f `
+                        $Port) +`
+                    '$Server.Start();' +`
+                    'New-Item -Type File /server-started.flag -Force > $null;' +`
+                    '$Connection = $Server.AcceptTcpClient();' +`
+                    '$StreamReader = New-Object System.IO.StreamReader($Connection.GetStream());' +`
+                    '$Input = $StreamReader.ReadLine();' +`
+                    (New-WaitFileCommand "/flows-tested.flag") +`
+                    '$StreamReader.Close();' +`
+                    '$Connection.Close();' +`
+                    '$Server.Stop();' +`
+                    'New-Item -Type File /server-stopped.flag -Force > $null;' +`
+                    'return $Input;' `
+                )
+
+                $Output = & docker exec $Container2Name powershell -Command $Command
+                return $Output
+            } -ArgumentList $Using:Port, $Using:Container2Name, $Using:Message
+        }
 
+        Write-Host "        Waiting for server-started"
+        Wait-RemoteFile -Session $Session2 -ContainerName $Container2Name -Path /server-started.flag
```

Comment:
Isn't there any other way to handle this case, without using temp files?

Reply:
The only other way I know about is to use carefully selected timeouts everywhere, which I don't think is a good idea, since there are multiple dependencies here:
* client waits for server
* flow test waits for client
* client and server wait for flow test completion to be able to disconnect
* another flow test waits for client and server to finish (this can be converted to `Wait-Job`)

The only case I can think of that doesn't require such synchronization is to only test outgoing flows (without even starting a server), but that's really weird scenario (but maybe we do really want that)